io-engine/src/bdev/nexus/nexus_channel.rs

```diff
@@ -77,26 +80,45 @@ impl<'n> NexusChannel<'n> {
     pub(crate) fn new(nexus: Pin<&mut Nexus<'n>>) -> Self {
         debug!("{nexus:?}: new channel on core {c}", c = Cores::current());
 
+        let b_init_thrd_hdls =
+            super::ENABLE_IO_ALL_THRD_NX_CHAN.load(Ordering::SeqCst);
+        let is_io_chan =
+            Thread::current().unwrap() != Thread::primary() || b_init_thrd_hdls;
+
         let mut writers = Vec::new();
         let mut readers = Vec::new();
 
-        nexus
-            .children_iter()
-            .filter(|c| c.is_healthy())
-            .for_each(|c| match (c.get_io_handle(), c.get_io_handle()) {
-                (Ok(w), Ok(r)) => {
-                    writers.push(w);
-                    readers.push(r);
-                }
-                _ => {
-                    c.set_faulted_state(FaultReason::CantOpen);
-                    error!(
-                        "Failed to get I/O handle for {c}, \
-                        skipping block device",
-                        c = c.uri()
-                    )
-                }
-            });
+        if is_io_chan {
+            nexus.children_iter().filter(|c| c.is_healthy()).for_each(
+                |c| match (c.get_io_handle(), c.get_io_handle()) {
+                    (Ok(w), Ok(r)) => {
+                        writers.push(w);
+                        readers.push(r);
+                    }
+                    _ => {
+                        c.set_faulted_state(FaultReason::CantOpen);
+                        error!(
+                            "Failed to get I/O handle for {c}, \
+                            skipping block device",
+                            c = c.uri()
+                        )
+                    }
+                },
+            );
+        } else {
+            // If we are here, this means the nexus channel being created is not
+            // the one to be used for normal IOs. Such a channel is
+            // created in rebuild path today, and it's on the init
+            // thread. The channels that we use for normal nexus IO
+            // are not on init thread however, those are on spdk threads
+            // created for nvmf target during poll group init. Those are the
+            // spdk threads named mayastor_nvmf_tcp_pg_core_*. The
+            // channel on init thread is only used for rebuild IOs.
+            // And the rebuild IOs are dispatched by
+            // directly calling write API without going via writers abstraction.
+            // Refer GTM-1075 for the race condition details.
+            debug!("{nexus:?}: skip nexus channel setup({}). is_io_channel: {is_io_chan}", Thread::current().unwrap().name());
```

Comment:
You could use Thread's debug formatter instead of name(). It prints core and thread name.
`debug!("{nexus:?}: skip nexus channel setup({t:?}). is_io_channel: {is_io_chan}", t = Thread::current().unwrap());`